### PR TITLE
Migrate Llama4DecoderLayer and Llama4ScannableBlock to NNX

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

* Migrate Llama4 text layers to NNX
* Remove unused `load_balance_loss`, which is currently enabled for dropping only, and llama4 only supports dropless ([here](https://github.com/AI-Hypercomputer/maxtext/blob/0f12f3da34607759aba4532dacf102c47c220215/src/MaxText/pyconfig.py#L338)).
* Updated the supported methods (thanks @shuningjin for the reminder)

# Tests

## Training

* Before the change - [logs](https://cloudlogging.app.goo.gl/ne95eXtTbTPivzbf8)
* After the change - [logs](https://cloudlogging.app.goo.gl/cv9BnJ6JCzfBgRS39)

```
# Before
Total memory size: 111.9 GB, Output size: 37.6 GB, Temp size: 74.3 GB, Argument size: 37.6 GB, Host temp size: 0.0 GB.
Using (GB) 38.15 / 95.74 (39.847504%) on TPU_8(process=0,(0,0,0,0))

completed step: 6, seconds: 13.701, TFLOP/s/device: 123.013, Tokens/s/device: 1195.841, total_weights: 262144, loss: 12.405

# After
Total memory size: 111.9 GB, Output size: 37.6 GB, Temp size: 74.3 GB, Argument size: 37.6 GB, Host temp size: 0.0 GB.
Using (GB) 38.15 / 95.74 (39.847504%) on TPU_0(process=0,(0,0,0,0))

completed step: 6, seconds: 13.686, TFLOP/s/device: 123.146, Tokens/s/device: 1197.138, total_weights: 262144, loss: 12.403
```

## Decoding

Noticed extra 1.2GB memory usage in `RAMstats: After load_params` 

* Before the change - [logs](https://cloudlogging.app.goo.gl/5Va7XG2GPCvU6TFV7)
* After the change - [logs](https://cloudlogging.app.goo.gl/PRhDenK7vbcqtvRc6)

```
Before

INFO 2025-10-01T23:23:05.733151482Z [resource.labels.containerName: jax-tpu] {}
INFO 2025-10-01T23:23:05.733193390Z [resource.labels.containerName: jax-tpu] Memstats: After load_params:
INFO 2025-10-01T23:23:05.733197370Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_4(process=1,(0,0,1,0))
INFO 2025-10-01T23:23:05.733199798Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_5(process=1,(1,0,1,0))
INFO 2025-10-01T23:23:05.733209563Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_6(process=1,(0,1,1,0))
INFO 2025-10-01T23:23:05.733213353Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_7(process=1,(1,1,1,0))

INFO 2025-10-01T23:23:05.733220215Z [resource.labels.containerName: jax-tpu] RAMstats: After load_params:
INFO 2025-10-01T23:23:05.733377282Z [resource.labels.containerName: jax-tpu] Using (GB) 15.63 / 440.84 (3.545504%) --> Available:422.44
INFO 2025-10-01T23:24:04.226796401Z [resource.labels.containerName: jax-tpu] Input `I love to` -> ` use the word “soul” in my writing. It’s a word that has a lot of meaning and can be used in a lot of different ways. I think it’s a word that can be used to describe a lot of different things, and I think it’s a word that can be used to describe a lot of different people.


After

INFO 2025-10-01T23:21:57.777933154Z [resource.labels.containerName: jax-tpu] Memstats: After load_params:
INFO 2025-10-01T23:21:57.777936309Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_4(process=1,(0,0,1,0))
INFO 2025-10-01T23:21:57.777943718Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_5(process=1,(1,0,1,0))
INFO 2025-10-01T23:21:57.777945827Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_6(process=1,(0,1,1,0))
INFO 2025-10-01T23:21:57.777947795Z [resource.labels.containerName: jax-tpu] Using (GB) 50.2 / 95.74 (52.433675%) on TPU_7(process=1,(1,1,1,0))

INFO 2025-10-01T23:21:57.777952490Z [resource.labels.containerName: jax-tpu] RAMstats: After load_params:
INFO 2025-10-01T23:21:57.778108970Z [resource.labels.containerName: jax-tpu] Using (GB) 16.82 / 440.84 (3.815443%) --> Available:421.26
INFO 2025-10-01T23:22:55.166921501Z [resource.labels.containerName: jax-tpu] Input `I love to` -> ` use the word “soul” in my writing. It’s a word that has a lot of meaning and can be used in a lot of different ways. I think it’s a word that can be used to describe a lot of different things, and I think it’s a word that can be used to describe a lot of different people.

```

## JetStream

Noticed 50GB less memory usage in RAMstats (from code this indicates the CPU memory usage)

* logs in diff - [link](https://diff.googleplex.com/#key=HTXNT73BVLpa) (no obvious diff)
* Profile matching - [link](https://screenshot.googleplex.com/5RsqazkUc5DNQGi)
* Profile before the change: [link](https://xprof.corp.google.com/memory_viewer/ranran-10261707462630720157) 
* Profile after the change: [link](https://xprof.corp.google.com/memory_profile/ranran-14079089929521082001)

```
# Before

Memstats: After load_params:
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_0(process=0,(0,0,0,0))
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_1(process=0,(1,0,0,0))
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_2(process=0,(0,1,0,0))
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_3(process=0,(1,1,0,0))

RAMstats: After load_params:
	Using (GB) 300.04 / 440.83 (68.062518%) -->  Available:138.08

# After

Memstats: After load_params:
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_0(process=0,(0,0,0,0))
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_1(process=0,(1,0,0,0))
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_2(process=0,(0,1,0,0))
	Using (GB) 50.19 / 95.74 (52.423230%) on TPU_3(process=0,(1,1,0,0))

RAMstats: After load_params:
	Using (GB) 247.21 / 440.83 (56.078307%) -->  Available:190.91
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/layers/llama4.py [hunk 1]
```diff
@@ -520,28 +531,47 @@ def __call__(
     else:
       return layer_output
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/MaxText/layers/llama4.py:544]: Added doc string to both `__init__` functions.

File: src/MaxText/layers/llama4.py [hunk 2]
```diff
@@ -445,66 +512,22 @@ def __call__(
...
-    layer_output = nn.Dropout(rate=cfg.dropout_rate, broadcast_dims=(-2,))(layer_output, deterministic=deterministic)
-
-    layer_output = nn.with_logical_constraint(
-        layer_output,
-        ("activation_batch", "activation_norm_length", "activation_embed"),
-    )
-
-    # NOTE: this is only needed for dropping MoE
-    if load_balance_loss is not None:
-      self.sow("intermediates", "moe_lb_loss", load_balance_loss)
```
Comment (bvandermoon) [inline src/MaxText/layers/llama4.py:507]: Why is this removed?

**Review by bvandermoon — APPROVED** [review]
Generally LGTM, left one question on a code snippet that was removed. Still not sure why the modules are not showing up, but doesn't seem related to the NNX migration. Your call on that since the HLOs that are there do match

File: src/MaxText/layers/llama4.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/layers/llama4.py:507]: Yeah, the load_balance_loss is always None, and also mentioned here: https://screenshot.googleplex.com/6dwFWhQwwkLmDot. Let me know if any questions :)
